src/pkgcheck/checks/overlays.py

```diff
@@ -132,3 +132,34 @@ def feed(self, pkg):
             non_local_use = pkg.iuse_stripped.difference(pkg.local_use.keys())
             if flags := self.unused_master_flags.intersection(non_local_use):
                 yield UnusedInMastersGlobalUse(sorted(flags), pkg=pkg)
+
+
+class MasterPackageClobbered(results.PackageResult, results.Error):
+    """Package in master is clobbered by package in overlay.
+
+    A repository (such as ``::guru``) that is supposed to only provide
+    additional packages to its masters (``::gentoo``), provides a package
+    that is found in its master, therefore potentially clobbering it.
+    """
+
+    def __init__(self, repository: str, **kwarg):
+        super().__init__(**kwarg)
+        self.repository = repository
+
+    @property
+    def desc(self):
+        return f"package from repository ::{self.repository} clobbered"
+
+
+class MasterPackageClobberedCheck(OverlayRepoCheck, OptionalCheck):
+    """Detect clobbering packages from master."""
+
+    _source = sources.PackageRepoSource
+    known_results = frozenset([MasterPackageClobbered])
+
+    def feed(self, pkgset):
+        for pkg in pkgset:
+            for repo in self.options.target_repo.masters:
+                if repo.has_match(pkg.unversioned_atom):
+                    yield MasterPackageClobbered(str(repo), pkg=pkg)
+            break
```

Comment:
```suggestion
        pkg = pkgset[0]
        for repo in self.options.target_repo.masters:
            if repo.has_match(pkg.unversioned_atom):
                yield MasterPackageClobbered(str(repo), pkg=pkg)
```

it won't enter here with empty `pkgset`